### ISSUE TITLE
Minio server stops in every few hours

### ISSUE THREAD

**Issue by puneetpandey37** [minio/minio#3565]
## Expected Behavior

Expected behavior is for Server to remain up till it is not stopped. 

## Current Behavior

Server stops(process dies) in every few hours. 

## Steps to Reproduce (for bugs)

1. docker pull minio/minio:edge
2. docker run -p 9000:9000 minio/minio:edge server /export
3. Wait for few hours and the Server would go down

## Context

Because of this issue, I need to restart the Server in every few hours.
Then I restart it and a new public and private key is generated. Then I need to change the keys in my code and redeploy it.  A new Server instance is created every time I restart it, which essentially means that all my files, stored in previous Server instance, get lost.

## Your Environment

* Server type and version: Latest from Master branch
* Operating System and version: centos-release-7-3.1611.el7.centos.x86_64

**Comment by puneetpandey37** [comment]
[dmesg.txt](https://github.com/minio/minio/files/699840/dmesg.txt)
[messages.txt](https://github.com/minio/minio/files/699841/messages.txt)

**Comment by harshavardhana** [comment]
To avoid the access keys being regenerated always set it through an env http://docs.minio.io/docs/minio-docker-quickstart-guide 

```
docker run -p 9000:9000 --name minio1 \
  -e "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE" \
  -e "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY" \
  -v /mnt/export/minio1:/export \
  -v /mnt/config/minio1:/root/.minio \
  minio/minio:edge server /export
```

**Comment by harshavardhana** [comment]
dmesg and messages don't provide any info. Would it be possible for you to capture 'docker logs'  from the container when the minio server stops?

This would also capture the server logs if any as well.  Can you also consistently reproduce this with a particular type of workload? if yes then please provide a sample code - we will replicate the environment.

**Comment by puneetpandey37** [comment]
Cannot provide you the source code as this is a big production ready code.
The server is again down. Will upload log files for you to examine.

**Comment by puneetpandey37** [comment]
I take the docker container log like this : "docker logs 19adbb9dc849"
19adbb9dc849 is the last container created. Below is the logs: 

Created minio configuration file at /root/.minio

Endpoint:  http://172.17.0.2:9000  http://127.0.0.1:9000
AccessKey: QPOHSC0AJ3H6QDZH0APC
SecretKey: LtrBD90HJNcJouvZg0+imsiw+3mBm93ZdplXq+cE
Region:    us-east-1
SQS ARNs:  <none>

Browser Access:
   http://172.17.0.2:9000  http://127.0.0.1:9000

Command-line Access: https://docs.minio.io/docs/minio-client-quickstart-guide
   $ mc config host add myminio http://172.17.0.2:9000 QPOHSC0AJ3H6QDZH0APC LtrBD90HJNcJouvZg0+imsiw+3mBm93ZdplXq+cE

Object API (Amazon S3 compatible):
   Go:         https://docs.minio.io/docs/golang-client-quickstart-guide
   Java:       https://docs.minio.io/docs/java-client-quickstart-guide
   Python:     https://docs.minio.io/docs/python-client-quickstart-guide
   JavaScript: https://docs.minio.io/docs/javascript-client-quickstart-guide

Drive Capacity: 20 GiB Free, 49 GiB Total

**Comment by harshavardhana** [comment]
> I take the docker container log like this : "docker logs 19adbb9dc849"
> 19adbb9dc849 is the last container created. Below is the logs:

Strange @puneetpandey37 - is it possible to remotely access and monitor the server? can send you a tool to inspect the situation here.. 

Can you run  `journalctl -u docker.service` and attach the output here? - Assuming you are on ubuntu - let me know if you are on a different operating system.

**Comment by harshavardhana** [comment]
As a workaround you can run `` --restart always``  to your docker command line this would restart minio if it silently crashes.

**Comment by puneetpandey37** [comment]
Mine is Centos --> centos-release-7-3.1611.el7.centos.x86_64

**Comment by puneetpandey37** [comment]
Did not get Restart command. Can you please give me the full command for that?

**Comment by puneetpandey37** [comment]
Jan 11 13:07:15 cloudtest systemd[1]: Starting Docker Application Container Engine...
Jan 11 13:07:15 cloudtest dockerd[18744]: time="2017-01-11T13:07:15.495254790+05:30" level=info msg="libcontainerd: new containerd process, pid: 18747"
Jan 11 13:07:16 cloudtest dockerd[18744]: time="2017-01-11T13:07:16.706998849+05:30" level=warning msg="devmapper: Usage of loopback devices is strongly discouraged for
Jan 11 13:07:17 cloudtest dockerd[18744]: time="2017-01-11T13:07:17.066912702+05:30" level=info msg="devmapper: Creating filesystem xfs on device docker-8:3-262466-base
Jan 11 13:07:17 cloudtest dockerd[18744]: time="2017-01-11T13:07:17.658439728+05:30" level=info msg="devmapper: Successfully created filesystem xfs on device docker-8:3
Jan 11 13:07:17 cloudtest dockerd[18744]: time="2017-01-11T13:07:17.786849876+05:30" level=info msg="Graph migration to content-addressability took 0.00 seconds"
Jan 11 13:07:17 cloudtest dockerd[18744]: time="2017-01-11T13:07:17.787336161+05:30" level=warning msg="mountpoint for pids not found"
Jan 11 13:07:17 cloudtest dockerd[18744]: time="2017-01-11T13:07:17.787705426+05:30" level=info msg="Loading containers: start."
Jan 11 13:07:17 cloudtest dockerd[18744]: time="2017-01-11T13:07:17.894684339+05:30" level=warning msg="Running modprobe bridge br_netfilter failed with message: modpro
Jan 11 13:07:18 cloudtest dockerd[18744]: time="2017-01-11T13:07:18.032520060+05:30" level=info msg="Firewalld running: false"
Jan 11 13:07:18 cloudtest dockerd[18744]: time="2017-01-11T13:07:18.467774300+05:30" level=info msg="Default bridge (docker0) is assigned with an IP address 172.17.0.0/
Jan 11 13:07:18 cloudtest dockerd[18744]: time="2017-01-11T13:07:18.786371030+05:30" level=info msg="Loading containers: done."
Jan 11 13:07:18 cloudtest dockerd[18744]: time="2017-01-11T13:07:18.786575673+05:30" level=info msg="Daemon has completed initialization"
Jan 11 13:07:18 cloudtest dockerd[18744]: time="2017-01-11T13:07:18.786601092+05:30" level=info msg="Docker daemon" commit=78d1802 graphdriver=devicemapper version=1.12
Jan 11 13:07:18 cloudtest dockerd[18744]: time="2017-01-11T13:07:18.794411087+05:30" level=info msg="API listen on /var/run/docker.sock"
Jan 11 13:07:18 cloudtest systemd[1]: Started Docker Application Container Engine.
Jan 11 13:07:30 cloudtest dockerd[18744]: time="2017-01-11T13:07:30.669507605+05:30" level=error msg="Handler for POST /v1.24/containers/create returned error: No such
Jan 11 13:07:40 cloudtest dockerd[18744]: time="2017-01-11T13:07:40.434739804+05:30" level=warning msg="failed to close stdin: rpc error: code = 2 desc = containerd: co
Jan 11 16:41:20 cloudtest dockerd[18744]: time="2017-01-11T16:41:20.696269295+05:30" level=warning msg="container kill failed because of 'container not found' or 'no su
Jan 11 21:30:08 cloudtest dockerd[18744]: time="2017-01-11T21:30:08.868260184+05:30" level=warning msg="container kill failed because of 'container not found' or 'no su

**Comment by harshavardhana** [comment]
On CentOS perhaps the file is `/var/log/daemon.log` @puneetpandey37 

```
docker run -p 9000:9000 --name minio1 \
  --restart always \ 
  -e "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE" \
  -e "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY" \
  -v /mnt/export/minio1:/export \
  -v /mnt/config/minio1:/root/.minio \
  minio/minio server /export
```

**Comment by puneetpandey37** [comment]
I am not able to start Server with this command.

I copy and paste this complete command and below is what I get:

[root@cloudtest log]# docker run -p 9000:9000 --name minio1 \
>   --restart always
"docker run" requires at least 1 argument(s).
See 'docker run --help'.

Usage:  docker run [OPTIONS] IMAGE [COMMAND] [ARG...]

Run a command in a new container
[root@cloudtest log]#   -e "MINIO_ACCESS_KEY=QPOHSC0AJ3H6QDZH0APC" \
>   -e "MINIO_SECRET_KEY=LtrBD90HJNcJouvZg0+imsiw+3mBm93ZdplXq+cE" \
>   -v /mnt/export/minio1:/export \
>   -v /mnt/config/minio1:/root/.minio \
>   minio/minio server /export
-bash: -e: command not found
[root@cloudtest log]# docker run -p 9000:9000 --name minio1 \
>   --restart always
"docker run" requires at least 1 argument(s).
See 'docker run --help'.

Usage:  docker run [OPTIONS] IMAGE [COMMAND] [ARG...]

**Comment by puneetpandey37** [comment]
And the daemon.log is not present in /var/log

**Comment by puneetpandey37** [comment]
I align the command and run:

[root@cloudtest log]# docker run -p 9000:9000 --name minio1 \ --restart always -e "MINIO_ACCESS_KEY=QPOHSC0AJ3H6QDZH0APC" \ -e "MINIO_SECRET_KEY=LtrBD90HJNcJouvZg0+imsiw+3mBm93ZdplXq+cE" \ -v /mnt/export/minio1:/export \ -v /mnt/config/minio1:/root/.minio \ minio/minio server /export



And this is what I get:

docker: Error parsing reference: " --restart" is not a valid repository/tag.
See 'docker run --help'.

**Comment by harshavardhana** [comment]
```
docker run -p 9000:9000 --name minio1 --restart always -e "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE" -e "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY" -v /mnt/export:/export  -v /mnt/config/minio1:/root/.minio minio/minio server /export

Endpoint:  http://172.17.0.2:9000  http://127.0.0.1:9000
AccessKey: AKIAIOSFODNN7EXAMPLE 
SecretKey: wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY 
Region:    us-east-1
SQS ARNs:  <none>

Browser Access:
   http://172.17.0.2:9000  http://127.0.0.1:9000

Command-line Access: https://docs.minio.io/docs/minio-client-quickstart-guide
   $ mc config host add myminio http://172.17.0.2:9000 AKIAIOSFODNN7EXAMPLE wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY

Object API (Amazon S3 compatible):
   Go:         https://docs.minio.io/docs/golang-client-quickstart-guide
   Java:       https://docs.minio.io/docs/java-client-quickstart-guide
   Python:     https://docs.minio.io/docs/python-client-quickstart-guide
   JavaScript: https://docs.minio.io/docs/javascript-client-quickstart-guide

Drive Capacity: 26 GiB Free, 55 GiB Total

``` 

This works my docker version is

```
docker -v
Docker version 1.12.1, build 23cf63
```

**Comment by puneetpandey37** [comment]
I have copied the one you pasted hare(in last comment of yours) and seem to go fine.
But it says that port is already allocated(even though I have stopped the server)

**Comment by harshavardhana** [comment]
> I have copied the one you pasted hare(in last comment of yours) and seem to go fine.
> But it says that port is already allocated(even though I have stopped the server)

It must be because you have another service running on port 9000.

**Comment by puneetpandey37** [comment]
I will take care of this. I will run the command(restart) you suggested. But please help me with the real issue i.e. Server going down in every few hours. It really worries me now. I took my management team in confidence and started this project to integrate Minio... In a day or two, I have to install this in one of our Private Cloud vendor's network and give a demo of it...

**Comment by puneetpandey37** [comment]
I used the command and the server is up again with same old keys... But the files are lost...
Is there a way to restart the Server and get both - old key and stored files???

And this is only a workaround anyway... We would need to make the server stable that never goes down by itself.

**Comment by harshavardhana** [comment]
> I used the command and the server is up again with same old keys... But the files are lost...
> Is there a way to restart the Server and get both - old key and stored files???

There is something strange going on here, we have https://play.minio.io:9000 running for months - so it would be better if we can get remote access and monitor it / observe. 

@puneetpandey37 also did you provide the proper path to the persistent data volume where you stored your files through docker command line?

**Comment by puneetpandey37** [comment]
Providing a remote access may not be possible coz I run the Server in my office VPN...

**Comment by harshavardhana** [comment]
> Providing a remote access may not be possible coz I run the Server in my office VPN...

Let me think and get back to you on debugging tools - perhaps a new binary with some more instrumentation. 

Can you paste your `docker run --rm minio/minio version` under use? can you try the `docker run --rm minio/minio:edge` and see if the issue is reproducible?

**Comment by donatello** [comment]
@puneetpandey37 To help solve this issue, could you please provide the following:

1. What is the command being used to the server?
2. What is the version of docker being used? I see that you are running on CentOS
3. What kind of requests are being made to the Minio server after it starts and before it crashes?
4. How much RAM is available on the system? And, are you restricting RAM usage in your docker command/container for the Minio server?

**Comment by puneetpandey37** [comment]
@harshavardhana I have removed all images and containers.. and again did the below:
docker pull minio/minio:edge
docker run -p 9000:9000 minio/minio:edge server /export

@donatello please find below information:

1- I use Docker to pull and run the Server.
docker pull minio/minio:edge
docker run -p 9000:9000 minio/minio:edge server /export

2- Docker version: Docker version 1.12.6, build 78d1802

3- All types of Requests - GET, PUT, POST Form, DELETE, Pre-signed URLs etc.. 
   Not sure about the request made before Server crashes. 

4- I do not explicitly restrict the RAM usage. Please find the RAM details below:

               
total:    1.8G        
used:    899M        
free:     109M        
shared:  5.1M        
buff/cache:  831M        
available: 765M

**Comment by donatello** [comment]
@puneetpandey37 I believe you may be running into an issue related to overlayfs used by default in docker when you do not mount a volume for minio (possibly related to this one - https://github.com/minio/minio/issues/3118 ). Use the alternative method described below.

In general, the way you have used is not the recommended way to run Minio for anything other than minor development. The better way is to use the method described on point 3 at https://docs.minio.io/docs/minio-docker-quickstart-guide:

```
docker run -p 9000:9000 --name minio1 \
  -v /mnt/export/minio1:/export \
  -v /mnt/config/minio1:/root/.minio \
  minio/minio server /export
```

You can read a bit more about docker volumes [here](https://docs.docker.com/engine/tutorials/dockervolumes/). 

The above command (`-v` option) mounts the `/mnt/export/minio1` into `/export` "inside" the container. Thus the minio server will store your files in this directory.

In your setup, to use the above command just make sure there is enough free disk space at `/mnt`. The data stored through Minio will be available at `/mnt/export/minio1` - the problem of losing your data after a restart should also not happen any more. The data will be stored in this directory.

The configuration for your Minio server will be available in the `/mnt/config/minio1` directory.

**Comment by donatello** [comment]
Additionally, as @harshavardhana described earlier in this thread, you should also set environment variables for access credentials, so they are not regenerated on a restart, for example:

```
docker run -p 9000:9000 --name minio1 \
  --restart always \ 
  -e "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE" \
  -e "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY" \
  -v /mnt/export/minio1:/export \
  -v /mnt/config/minio1:/root/.minio \
  minio/minio server /export

```

**Comment by donatello** [comment]
@puneetpandey37 When running the commands given above, please be sure copy paste the commands exactly - the earlier problems of getting errors and different output seems to be because you are not carefully copy-pasting the commands.

**Comment by harshavardhana** [comment]
@puneetpandey37 are you still seeing this issu ? - if not would you be interested some more instrumentation to help us debug this? 

Do let us know.

**Comment by puneetpandey37** [comment]
I stopped and deleted all other containers and images(there ware many). And have started the server again. Will let you guys know when it crashes.

**Comment by puneetpandey37** [comment]
And the Server goes down again. It's been happening all these days.

I start the Server with the command you guys suggested:

docker run -p 9000:9000 --name minio1 \
  --restart always \ 
  -e "MINIO_ACCESS_KEY=AKIAIOSFODNN7EXAMPLE" \
  -e "MINIO_SECRET_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY" \
  -v /mnt/export/minio1:/export \
  -v /mnt/config/minio1:/root/.minio \
  minio/minio server /export

**Comment by puneetpandey37** [comment]
![image](https://cloud.githubusercontent.com/assets/5543251/22012785/de5952b8-dcbb-11e6-8d9a-e3544f970db8.png)

**Comment by harshavardhana** [comment]
@puneetpandey37 can you do `docker logs <container_id>`

**Comment by harshavardhana** [comment]
and provide us the output?

**Comment by puneetpandey37** [comment]
Sorry! I just deleted the Image and Container! I am (very carefully) starting it all over again. I will put here every single step I take to bring up the Server. And will put here the container log when the server crashes.

**Comment by harshavardhana** [comment]
> Sorry! I just deleted the Image and Container! I am (very carefully) starting it all over again. I will put here every single step I take to bring up the Server. And will put here the container log when the server crashes.

Okay understood @puneetpandey37

**Comment by puneetpandey37** [comment]
No Docker Container and Image exists:
![image](https://cloud.githubusercontent.com/assets/5543251/22014319/209d0460-dcc3-11e6-9419-194c244104b2.png)


Now I start the Server by pulling from **edge**:

![image](https://cloud.githubusercontent.com/assets/5543251/22014360/54a25346-dcc3-11e6-9ae1-a5e87527ed7a.png)

Below is the Server Console logs for Server Startup:

![image](https://cloud.githubusercontent.com/assets/5543251/22014407/87309732-dcc3-11e6-9a06-8a782be786a1.png)


Will provide you the Container log when Server crashes.

**Comment by puneetpandey37** [comment]
Exited 10 minutes back:

![image](https://cloud.githubusercontent.com/assets/5543251/22025306/2016dbda-dcf3-11e6-9198-7c8904aae82f.png)

Please find the container logs below:

![image](https://cloud.githubusercontent.com/assets/5543251/22025349/47be40ec-dcf3-11e6-805f-e0b569f3c0b7.png)

**Comment by vadmeste** [comment]
@puneetpandey37, log that we see now is just verbose and not important, can you manage to paste all log instead ?

**Comment by puneetpandey37** [comment]
This is all I get by this: docker logs bb92d5dd9fef

**Comment by puneetpandey37** [comment]
![image](https://cloud.githubusercontent.com/assets/5543251/22026179/103a7e80-dcf6-11e6-86ed-54a5c5f9e5f3.png)


The above log is from /var/lib/docker/containers/bb92d5dd9fefca2f7004c4b1e894181c314b4c13feb4231aecef60bc2663183d

bb92d5dd9fefca2f7004c4b1e894181c314b4c13feb4231aecef60bc2663183d-json.log

**Comment by vadmeste** [comment]
Thanks @puneetpandey37, I read this thread but can you paste again the contents of dmesg ? You have only 2Gb of total RAM and I suspect Minio is killed by your kernel.

**Comment by puneetpandey37** [comment]
Last few lines of dmesg:

![image](https://cloud.githubusercontent.com/assets/5543251/22026858/01895f30-dcf8-11e6-96dc-61e975a49a9a.png)

**Comment by puneetpandey37** [comment]
And FYI, the same machine hosts and runs very heavy CAE engines like: Geometry, VPS, OpenFOAM, and Procast.

**Comment by vadmeste** [comment]
@puneetpandey37, what's the output of `dmesg | grep "Out of memory"` ?

**Comment by puneetpandey37** [comment]
[3906003.769571] Out of memory: Kill process 21003 (MesherEngine) score 662 or sacrifice child

**Comment by puneetpandey37** [comment]
dmesg | grep "kill"

![image](https://cloud.githubusercontent.com/assets/5543251/22027459/f9a7adce-dcf9-11e6-80d5-fbc9b02de970.png)

**Comment by harshavardhana** [comment]
you should paste the entire trace before this actually occurred @puneetpandey37 can you attach the 'dmesg' output? 

From what it looks like the likely cause `MesherEngine` OOM is taking multiple processes down by kernel.  The algorithm for OOM kill is based on `oom_score` and it is quite possible that is is consistently happening because `MesherEngine` uses more memory disallowing or killing other process from allocating new memory.

**Comment by puneetpandey37** [comment]
I again started the server. This time it was crashed in few minutes only. Please find the attached dmesg file.

[dmesg.txt](https://github.com/minio/minio/files/713082/dmesg.txt)


Meanwhile, I will install Minio in a separate machine and monitor it.

**Comment by harshavardhana** [comment]
dmesg doesn't have anything other than OOM kill for MesherEngine which we analysed before. Minio being run to a separate machine is better to isolate this problem. Please let us know how it goes.

**Comment by puneetpandey37** [comment]
And it goes down again. Even in a new machine.

![image](https://cloud.githubusercontent.com/assets/5543251/22061502/347ee54c-dd9b-11e6-9b66-3b3cc3a8497d.png)


No result for: dmesg | grep "Out of memory"

[dmesg.txt](https://github.com/minio/minio/files/713680/dmesg.txt)

**Comment by puneetpandey37** [comment]
New machine detail:

![image](https://cloud.githubusercontent.com/assets/5543251/22062740/41c66b66-dda1-11e6-9e4d-bbb9219b9dfe.png)

**Comment by puneetpandey37** [comment]
To me, it looks like a Bug now. I have provided the machine details and steps/commands I use to set up the server. Could you please try to replicate the issue.

**Comment by puneetpandey37** [comment]
From dmesg:

This is what it says when the Server is brought up:

**device vethaac76fd entered promiscuous mode
vethaac76fd: link is not ready
vethaac76fd: link becomes ready
docker0: port 1(vethaac76fd) entered forwarding state**


And this when it crashes:

**docker0: port 1(vethaac76fd) entered disabled state
device vethaac76fd left promiscuous mode
docker0: port 1(vethaac76fd) entered disabled state**

**Comment by puneetpandey37** [comment]
Just discovered, restarting the Docker brings up the Minio Server.

`sudo service docker restart`

![image](https://cloud.githubusercontent.com/assets/5543251/22072120/487163fa-ddc7-11e6-9dae-cb7ee25422e3.png)

@harshavardhana  @vadmeste

**Comment by puneetpandey37** [comment]
The problem may have something to do with WICD(wired and wireless network manager)... I have disabled wcid... Let us see if the problem reappears! @harshavardhana @vadmeste

**Comment by harshavardhana** [comment]
Working on an instrumentation to gather more info @puneetpandey37 - will ping you with a new docker link  for testing - Thanks for all the hardwork here.

**Comment by puneetpandey37** [comment]
FYI, the issue seems to be fixed after disabling the network manager. I will put here the exact steps I took to resolve the issue(may help someone, facing this issue):

1- Restarted the Docker and checked if the container got the port attached with it. And it got attached. 
2- Check if you have another NetworkManager installed. For me, issue was related to wicd (wired and wireless network manager) another network manager installed on my machine.
3- I had stopped WCD:
    `sudo service wicd stop`
4- And restarted the Docker:
   `sudo service docker restart`

With this change, issue seems to have resolved as the Minio is still up for last 9 hours(It would usually crash in every 2-3 hours)

![image](https://cloud.githubusercontent.com/assets/5543251/22091994/a0b79a6c-de20-11e6-8af3-a07d075575d4.png)


I will still monitor the server today and will close this ticket if things still look good. Thanks for all your help guys.  @harshavardhana @vadmeste
